pyro/contrib/gp/__init__.py

```diff
@@ -1,15 +1,136 @@
 from __future__ import absolute_import, division, print_function
 
+from torch.autograd import Variable
 import torch.nn as nn
-from torch.nn import Parameter
+from torch.distributions import transform_to
 
+import pyro
+import pyro.distributions as dist
 
-class InducingPoints(nn.Module):
 
-    def __init__(self, Xu, name="inducing_points"):
-        super(InducingPoints, self).__init__()
-        self.inducing_points = Parameter(Xu)
-        self.name = name
+class Parameterized(nn.Module):
+    """
+    Parameterized class.
 
-    def forward(self):
-        return self.inducing_points
+    This is a base class for other classes in Gaussian Process.
+    By default, a parameter will be a ``torch.nn.Parameter`` containing ``torch.FloatTensor``.
+    To cast them to the correct data type or GPU device, we can call methods such as
+    ``.double()``, ``.cuda(device=0)``,...
+    See `torch.nn.Module
+    <http://pytorch.org/docs/master/nn.html#torch.nn.Module>`_ for more information.
+    """
+
+    def __init__(self):
+        super(Parameterized, self).__init__()
+        self._priors = {}
+        self._constraints = {}
+        self._name = None
+        self._registered_params = {}
+        self._fixed_params = {}
+
+    def set_prior(self, param, prior):
+        """
+        Sets a prior to a parameter. Use `set_prior(param, None)` to remove the prior distribution of
+        the parameter.
+
+        :param str param: Name of a parameter.
+        :param pyro.distributions.Distribution prior: A prior distribution for random variable ``param``.
+        """
+        self._priors[param] = prior
+
+    def set_constraint(self, param, constraint):
+        """
+        Sets a constraint to a parameter. Use `set_constraint(param, None)` to remove the constraint of
+        the parameter.
+
+        :param str param: Name of a parameter.
+        :param torch.distributions.constraints.Constraint constraint: A Pytorch constraint.
+            See `Pytorch's docs
+            <http://pytorch.org/docs/master/distributions.html#module-torch.distributions.constraints>`_
+            for a list of constraints.
+        """
+        self._constraints[param] = constraint
+
+    def set_mode(self, mode):
+        """
+        Sets mode for the module to be able to use its parameters in stochastic functions.
+
+        :param str mode: Either "model" or "guide".
+        """
+        if mode not in ["model", "guide"]:
+            raise ValueError("Mode should be either 'model' or 'guide', but got {}.".format(mode))
+        for param in self._parameters:
+            self._register_param(param, mode)
+        return self
```

Comment:
nit: It's kind of weird to `return self` here. I realize some codebases use method chaining heavily, but the patter is seldom used in Pyro.

Reply:
I agree, it just provides me a shortcut for `kernel = self.kernel.set_mode(...)`, which is unnecessary.